cmd/certgen.go

```diff
@@ -228,6 +252,104 @@ func generateCertificates() error {
+			if err != nil {
+				return fmt.Errorf("failed to load ExternalWorkload CA: %w", err)
+			}
+		} else {
+			// Does the configmap already exist?
+			ctx, cancel := context.WithTimeout(context.Background(), option.Config.K8sRequestTimeout)
+			defer cancel()
+			err = externalworkloadCA.LoadFromConfigMap(ctx, k8sClient)
+			if err != nil {
+				log.WithError(err).Info("failed to load ExternalWorkload configmap CA, generating new CA")
+				err = externalworkloadCA.Generate(option.Config.ExternalWorkloadCACertCommonName, option.Config.ExternalWorkloadCACertValidityDuration)
+				if err != nil {
+					return fmt.Errorf("failed to generate ExternalWorkload CA: %w", err)
+				}
+			} else {
+				log.Info("Loaded ExternalWorkload CA ConfigMap")
+				haveCAConfigMap = true
+			}
+		}
+
+		log.Info("Generating server certificate for ExternalWorkload")
+		externalworkloadServerCert := generate.NewCert(
+			option.Config.ExternalWorkloadServerCertCommonName,
+			option.Config.ExternalWorkloadServerCertValidityDuration,
+			defaults.ExternalWorkloadCertUsage,
+			option.Config.ExternalWorkloadServerCertSecretName,
+			option.Config.CiliumNamespace,
+		).WithHosts([]string{option.Config.ExternalWorkloadServerCertCommonName, "127.0.0.1"})
+		err = externalworkloadServerCert.Generate(externalworkloadCA.CACert, externalworkloadCA.CAKey)
+		if err != nil {
+			return fmt.Errorf("failed to generate ExternalWorkload server cert: %w", err)
+		}
+
+		log.Info("Generating admin certificate for ExternalWorkload")
+		externalworkloadAdminCert := generate.NewCert(
+			option.Config.ExternalWorkloadAdminCertCommonName,
+			option.Config.ExternalWorkloadAdminCertValidityDuration,
+			defaults.ExternalWorkloadCertUsage,
+			option.Config.ExternalWorkloadAdminCertSecretName,
+			option.Config.CiliumNamespace,
+		).WithHosts([]string{"localhost"})
+		err = externalworkloadAdminCert.Generate(externalworkloadCA.CACert, externalworkloadCA.CAKey)
+		if err != nil {
+			return fmt.Errorf("failed to generate ExternalWorkload admin cert: %w", err)
+		}
+
+		log.Info("Generating client certificate for ExternalWorkload")
+		externalworkloadClientCert := generate.NewCert(
+			option.Config.ExternalWorkloadClientCertCommonName,
+			option.Config.ExternalWorkloadClientCertValidityDuration,
+			defaults.ExternalWorkloadCertUsage,
+			option.Config.ExternalWorkloadClientCertSecretName,
+			option.Config.CiliumNamespace,
+		)
+		err = externalworkloadClientCert.Generate(externalworkloadCA.CACert, externalworkloadCA.CAKey)
+		if err != nil {
+			return fmt.Errorf("failed to generate ExternalWorkload client cert: %w", err)
+		}
+
+		// Store the generated certs
+		if !haveCAConfigMap {
+			ctx, cancel := context.WithTimeout(context.Background(), option.Config.K8sRequestTimeout)
+			defer cancel()
+			if err := externalworkloadCA.StoreAsConfigMapWithKey(ctx, k8sClient); err != nil {
+				return fmt.Errorf("failed to create configmap for ExternalWorkload CA: %w", err)
+			}
+		}
+		{
```

Comment:
I assume these are here to avoid accidental shadowing of the `ctx, cancel` vars? The `defer cancel()` will only be executed at the end of the function in Go, so these won't do anything for the defer (which is fine).

Reply:
OK, was thinking too fast, or not enough, as I had the idea that defers would have been called on the end of the block. I'll simplify this as well.